File: crates/cumulus-core/src/maven.rs
use std::fs;
use std::path::Path;

pub fn get_maven_goals(pom_path: &Path) -> Vec<String> {
    let mut goals = vec![
        "clean".to_string(),
        "validate".to_string(),
        "compile".to_string(),
        "test".to_string(),
        "package".to_string(),
        "verify".to_string(),
        "install".to_string(),
        "deploy".to_string(),
        "clean compile".to_string(),
        "clean test".to_string(),
        "clean package".to_string(),
        "clean install".to_string(),
        "clean verify".to_string(),
    ];

    let content = match fs::read_to_string(pom_path) {
        Ok(c) => c,
        Err(_) => return goals,
    };

    let mut plugin_goals = Vec::new();

    if content.contains("spring-boot-maven-plugin") {
        plugin_goals.push("spring-boot:run".to_string());
        plugin_goals.push("spring-boot:build-image".to_string());
        plugin_goals.push("spring-boot:repackage".to_string());
    }
    if content.contains("quarkus-maven-plugin") {
        plugin_goals.push("quarkus:dev".to_string());
        plugin_goals.push("quarkus:build".to_string());
        plugin_goals.push("quarkus:test".to_string());
    }
    if content.contains("maven-surefire-plugin") || content.contains("<test") {
        plugin_goals.push("test-compile".to_string());
        plugin_goals.push("surefire:test".to_string());
    }
    if content.contains("maven-failsafe-plugin") {
        plugin_goals.push("failsafe:integration-test".to_string());
        plugin_goals.push("verify".to_string());
    }
    if content.contains("exec-maven-plugin") {
        plugin_goals.push("exec:java".to_string());
        plugin_goals.push("exec:exec".to_string());
    }

    plugin_goals.push("dependency:tree".to_string());
    plugin_goals.push("dependency:analyze".to_string());
    plugin_goals.push("dependency:resolve".to_string());
    plugin_goals.push("help:effective-pom".to_string());
    plugin_goals.push("help:active-profiles".to_string());

    for goal in plugin_goals {
        if !goals.contains(&goal) {
            goals.push(goal);
        }
    }

    goals
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;

    #[test]
    fn test_maven_goals_parsing() {
        let dir = tempfile::tempdir().unwrap();
        let pom_file = dir.path().join("pom.xml");
        let mut file = fs::File::create(&pom_file).unwrap();
        writeln!(file, "<project><build><plugins><plugin><artifactId>spring-boot-maven-plugin</artifactId></plugin></plugins></build></project>").unwrap();

        let goals = get_maven_goals(&pom_file);
        assert!(goals.contains(&"spring-boot:run".to_string()));
    }
}

File: lua/cumulus/util/maven.lua
-- Cumulus Maven Build Integration
--
-- Architecture: Lua is a bridge only. All parsing and analysis is done by
-- the cumulus-engine Scala binary. This file handles Neovim UI/terminal wiring.

local M = {}

M.offline_mode = false

function M.toggle_offline_mode()
  M.offline_mode = not M.offline_mode
  vim.notify(
    "Maven Offline Mode: " .. (M.offline_mode and "ENABLED (-o)" or "DISABLED"),
    vim.log.levels.INFO
  )
end

function M.find_pom(buf)
  local dir = (buf and vim.api.nvim_buf_is_valid(buf) and vim.fs.dirname(vim.api.nvim_buf_get_name(buf)))
    or vim.fn.expand("%:p:h")
  if dir == "" or not dir then
    dir = vim.fn.getcwd()
  end

  local engine = require("cumulus.util.engine")
  if engine.is_available() then
    local topo = engine.classify_workspace(dir)
    if topo then
      if topo.primary_type == "maven" then
        return true
      end
      if topo.project_types then
        for _, ptype in ipairs(topo.project_types) do
          if ptype == "maven" then
            return true
          end
        end
      end
    end
    local res = engine.discover_build_tool(dir) or engine.discover_build_tool(vim.fn.getcwd())
    if res and (res.tool == "maven" or res.build_tool == "maven") then
      return true
    end
  end

  local patterns = { "pom.xml", "mvnw" }
  return vim.fs.root(dir, patterns) ~= nil or vim.fs.root(vim.fn.getcwd(), patterns) ~= nil
end

function M.get_mvn_cmd()
  local cwd = vim.fn.getcwd()
  local mvnw = cwd .. "/mvnw"
  local base = "mvn"
  if vim.fn.filereadable(mvnw) == 1 then
    if vim.fn.executable(mvnw) == 0 then
      vim.fn.system({ "chmod", "+x", mvnw })
    end
    base = "./mvnw"
  end

  local engine = require("cumulus.util.engine")
  if M.offline_mode or (engine.is_available() and engine.check_network() == false) then
    return base .. " -o"
  end

  return base
end

function M.run_maven_cmd(cmd)
  if not M.find_pom() then
    require("cumulus.util.engine").notify_warn("No pom.xml found in project", "Cumulus JVM")
    return
  end

  local base_cmd = M.get_mvn_cmd()
  if cmd:sub(1, 4) == "mvn " then
    cmd = base_cmd .. cmd:sub(4)
  end

  require("cumulus.util.engine").run_term(cmd, { title = "Cumulus Maven" })
end

function M.sync_dependencies()
  if not M.find_pom() then
    return
  end

  local base_cmd = M.get_mvn_cmd()
  require("cumulus.util.sync-runner").run({
    cmd = { base_cmd, "-q", "dependency:resolve" },
    notify_id = "cumulus_maven_sync",
    tool_label = "Maven",
    base_cmd = base_cmd,
  })
end

--- Get Maven goals for the current project.
--- All pom.xml parsing is done by the cumulus-engine Scala binary (parse-pom subcommand).
function M.get_maven_goals()
  local cwd = vim.fn.getcwd()
  local pom_path = vim.fn.findfile("pom.xml", cwd .. ";")
  if pom_path == "" then
    local current_file = vim.fn.expand("%:p:h")
    if current_file ~= "" then
      pom_path = vim.fn.findfile("pom.xml", current_file .. ";")
    end
  end

  if pom_path == "" then
    return {}
  end

  local engine = require("cumulus.util.engine")
  if engine.is_available() then
    local goals = engine.parse_pom_goals(vim.fn.fnamemodify(pom_path, ":p"))
    if goals and #goals > 0 then
      return goals
    end
  end

  -- Fallback default lifecycle and common plugin goals
  local default_goals = {
    "clean",
    "validate",
    "compile",
    "test",
    "package",
    "verify",
    "install",
    "deploy",
    "clean compile",
    "clean test",
    "clean package",
    "clean install",
    "spring-boot:run",
    "quarkus:dev",
  }
  return default_goals
end

function M.run_maven_goal()
  if not M.find_pom() then
    vim.notify("No pom.xml found in project", vim.log.levels.WARN)
    return
  end

  vim.notify("Loading Maven goals...", vim.log.levels.INFO)
  local goals = M.get_maven_goals()
  local base_cmd = M.get_mvn_cmd()

  vim.ui.select(goals, {
    prompt = "Select Maven Goal:",
    format_item = function(item)
      return base_cmd .. " " .. item
    end,
  }, function(choice)
    if choice then
      M.run_maven_cmd(base_cmd .. " " .. choice)
    end
  end)
end

--- Get direct Maven project dependencies via engine helper.
---@param pom_path? string
---@return table[]|nil
function M.get_dependencies(pom_path)
  pom_path = pom_path or (vim.fn.getcwd() .. "/pom.xml")
  local engine = require("cumulus.util.engine")
  if engine.is_available() then
    return engine.resolve_deps(pom_path)
  end
  return nil
end

return M
